warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ValidateFilterFunctionVisitor.java

```diff
@@ -0,0 +1,412 @@
+import org.apache.commons.jexl2.parser.ASTFloatLiteral;
+import org.apache.commons.jexl2.parser.ASTForeachStatement;
+import org.apache.commons.jexl2.parser.ASTFunctionNode;
+import org.apache.commons.jexl2.parser.ASTGENode;
+import org.apache.commons.jexl2.parser.ASTGTNode;
+import org.apache.commons.jexl2.parser.ASTIdentifier;
+import org.apache.commons.jexl2.parser.ASTIfStatement;
+import org.apache.commons.jexl2.parser.ASTIntegerLiteral;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ASTLENode;
+import org.apache.commons.jexl2.parser.ASTLTNode;
+import org.apache.commons.jexl2.parser.ASTMapEntry;
+import org.apache.commons.jexl2.parser.ASTMapLiteral;
+import org.apache.commons.jexl2.parser.ASTMethodNode;
+import org.apache.commons.jexl2.parser.ASTModNode;
+import org.apache.commons.jexl2.parser.ASTMulNode;
+import org.apache.commons.jexl2.parser.ASTNENode;
+import org.apache.commons.jexl2.parser.ASTNRNode;
+import org.apache.commons.jexl2.parser.ASTNotNode;
+import org.apache.commons.jexl2.parser.ASTNullLiteral;
+import org.apache.commons.jexl2.parser.ASTNumberLiteral;
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.ASTReferenceExpression;
+import org.apache.commons.jexl2.parser.ASTReturnStatement;
+import org.apache.commons.jexl2.parser.ASTSizeFunction;
+import org.apache.commons.jexl2.parser.ASTSizeMethod;
+import org.apache.commons.jexl2.parser.ASTStringLiteral;
+import org.apache.commons.jexl2.parser.ASTTernaryNode;
+import org.apache.commons.jexl2.parser.ASTTrueNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTVar;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.SimpleNode;
+import org.apache.log4j.Logger;
+
+import java.util.Set;
+
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctions.EVAL_PHASE_FUNCTION_NAMESPACE;
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctionsDescriptor.INCLUDE_TEXT;
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctionsDescriptor.NO_EXPANSION;
+
+/**
+ * Verifies that no filter function is run against an index-only field
+ * <p>
+ * The exceptions to this rule are the #NO_EXPANSION and #TEXT functions.
+ * <p>
+ * The #NO_EXPANSION term will be pruned from the query prior to this visitor's execution, and the #TEXT query will have an index query added into the query.
+ */
+public class ValidateFilterFunctionVisitor extends BaseVisitor {
+    
+    private static final Logger log = ThreadConfigurableLogger.getLogger(ValidateFilterFunctionVisitor.class);
+    
+    private final Set<String> indexOnlyFields;
+    private final Set<String> excludedFunctions = Sets.newHashSet(NO_EXPANSION, INCLUDE_TEXT);
```

Comment:
This can be a static final immutable set instantiated via `Collections.unmodifiableSet()`. Additionally, I'd like to see a comment indicating that these are filter functions that support the usage of index-only fields to make it clear why these are excluded from the validation.

Reply:
I'll make the set immutable and add comments. Also created https://github.com/NationalSecurityAgency/datawave/issues/1615 to move NOEXPANSION to query functions

Reply:
These functions are excluded because they are NOT actually supposed to be filter functions. The class level docs hint at why, but I can make it explicit for why these functions are excluded from validation. The short summary is
- #INCLUDE is explicitly allowed to function against index only fields
- #NOEXPANSION controls query model expansion, to include index only fields

Reply:
Ah, gotcha, thanks for clarifying.